Rebased 'darkMode' onto 'master'.

gman-comicdb opened the pull request:
This brings the darkMode branch up-to-date with master branch.

gman-comicdb commented:
Thanks for resolving my errors - first pull. Still learning the GitHub ropes, like when to tell you have a new message :)
I'll merge in future and securitize the diff more closely.

I've been following https://github.com/maforget/ComicRackCE/issues/2; I agree that would be the easiest way but it's dependant on actually getting to a modern .NET version, which is very not-easy.

Dark Mode is a personal high-priority so that's where I'm focussing. As a C# beginner it's also a good way to learn the language and the CRCE codebase. If it can be done in an acceptable way, it would be great to have it as part of the main release.

maforget commented:
Well the merge wasn't done without problems. A lot of files have moved since then and it didn't catch all the files being moved making it so that there was missing files. Changed some config to max the rename search and had to use the command line or Github Desktop because Visual Studio wouldn't catch the renames.

In this kind of situation a rebase seems the best way to go, but a rebase rewrites history and when a fork is already in multiple repos changing the history can be chaos. It would have probably be ok since it wouldn't have probably not changed what was already in the various fork, but didn't want to take a chance. You can test all you want locally, but the moment it is in a remote it can be problematic.

You will probably have to do a force pull since my darkMode and yours differ, which is why when you do PR you are better creating a branch based on the source and work on that.

If you want to work on it go ahead. But I wanted to keep the changes to a minimum. Without having to add checks for the dark mode setting in various place and keep it all centralized. It will probably not be possible to do without adding checks in some places. There are a lot of layers that would require changing the underlying control and isn't just changing a value here and there. I believe some controls (like ComboBox, ProgressBar) will need to be created or updated to support dark mode. 

If you can make it not look like shit, good for you, it is a good learning exercise, although this code base is big and might be confusing for a beginner. But it is well done so it is a good reference.

gman-comicdb commented:
Damn... sorry for making you do all that work. Yes, I thought a rebase would be easier than a merge based on non-git source control I worked with in the past, and there being so many changes.

I've been pretty ~~lazy~~ careful about keeping most of my work local, and I was bearing in mind its a branch that probably no one is really working on. It'll be a while and considerably more checks/approvals before thinking about a mainline branch.

---

I've pushed my test Dark Mode feature out: [ComicRackCE/darkMode/feature-test](https://github.com/gman-comicdb/ComicRackCE/tree/darkMode/feature-test)
There's still some work to do and I need to write some notes under the issue, but I consider the bulk of it done.

### Approach
<sub>aka long version - idk if git has the concept of "child features", but the workflow I know that's what these are </sub>
- [ColorScheme](https://github.com/gman-comicdb/ComicRackCE/tree/darkMode/ColorScheme) - Extend `ColorScheme` as much as possible (... perhaps too much; would have preferring not to use event handlers)
- [uxtheme](https://github.com/gman-comicdb/ComicRackCE/tree/darkMode/uxtheme) - Add `UXTheme`, for actual theming using [uxtheme.dll](https://learn.microsoft.com/en-gb/windows/win32/api/uxtheme/)
- [formEx](https://github.com/gman-comicdb/ComicRackCE/tree/darkMode/formEx) - `Form` -> `FormEx`, which calls `ColorScheme` and `UXTheme` in `OnHandleCreated`
- [controlsEx](https://github.com/gman-comicdb/ComicRackCE/tree/darkMode/controlsEx) - Also added `ColorScheme` and `UXTheme` to extended control `OnHandleCreated` (ComboBoxEx etc)
- [modify-hardcoded-colors](https://github.com/gman-comicdb/ComicRackCE/tree/darkMode/modify-hardcoded-colors) - Started on the hardcoded stuff. Just CollapsibleGroupBox because small change, big effect.

This is a test feature; I'm not a C# native; it could be doing bad things without me realising; tried to make sure that's not the case.

### Still to-do
<sub>(or done but need to clean-up + push)</sub>
- Handle `CheckedListBox` and `RickTextBox`
- Hardcoded colors
- Resources (images)
- Test
    - change DPI
    - different OS versions

**Approach notes**
- Sadly I couldn't find a good way around having forms inherit from a class which calls the theming methods. Worse alternatives would have been to add the calls to all forms, which seems pointless when having a base class does the same thing afaik, or listening for forms being created, which didn't strike me as an elegant or safe method.
- UXTheme-ing is nice but v incomplete on my Win10 version; I think it'll work a lot better with Win11. It's actually already used in the code, in [one place](https://github.com/maforget/ComicRackCE/blob/9231547abd8ee73e7f6739be0ac02a07d249179a/cYo.Common.Windows/Forms/TanColorTable.cs#L60) - `TanColorTable` looks like a precursor to ColorScheme. I'd like to see if overriding `KnownColors` would get us further without having to tweak so many things, but I couldn't find good documentation and I cba to work out all the theme colors - doing it for a couple of elements was enough.
- Lots of pruning, reviewing + testing required

maforget commented:
Seems nice. I've only checked briefly and didn't test but here some notes. I am still not certain 100% about what everything is doing. Changing the form base is pretty much unavoidable, it make sense).

You seem to be coming from a C/C++ background based on some way of doing so. I am not familiar with it and my knowledge of using Native methods is basic at best.

- Is there a way why you are loading the uxtheme library using LoadLibraryEx + GetProcAddress. From what I understand you are basically doing what the framework is doing underneath. I believe you could just call the functions directly and wrap them is a catch for [EntryPointNotFoundException](https://learn.microsoft.com/en-us/dotnet/api/system.entrypointnotfoundexception?view=netframework-4.8) & [DllNotFoundException](https://learn.microsoft.com/en-us/dotnet/api/system.dllnotfoundexception?view=netframework-4.8). <br/><br/>Here is just the first example I got from searching Github for uxtheme.
https://github.com/Rectify11/Installer/blob/8f7903bdf63c91381e588c58e072da9050d01c16/Rectify11Installer/Win32/DarkMode.cs#L10-L15

- If possible when using constant like you calling 5644 on `SendMessage`, if we can use a name constant so it is easy what it does. I haven't looked it up, but it is easier to know what is going on. I am in the process of changing a lot of these values that might have came from the de compilation process.

- We should avoid hard coding values, even if it's for ease of refactoring. Things like your DarkColors Dictionary should probably be using a Color Enum instead of a string. When you call it in the `SetDarkMode` couldn't you just pass the Color directly than the name instead?

- Same with hard coding button names like `btnVTagsHelp` or `btAssociateExtensions` directly in an helper file. I like the idea to centralizing everything, but calling controls in a string should happen. Either use the control itself via a variable or delegate or use `nameof()` instead for specific classes. If it needs to be done for specific controls and can't be generic, then we should call the function from the dialog or set a variable or pass delegate with a list of controls and set the values in the `PreferencesDialog` and gets invoked by the ColorScheme. I want to have all values to tweak in a central place (especially if we want to add user modifiable values) but really don't want a hard coded button name. The moment we refactor or move things it will break and it will prevent analyzing the code if a certain control/class is referenced anywhere.

- I don't know what your idea is for images, the reference I used (and what you seem to have added) is just invert the colors. It might give weird results. we probably could just add alternative images that automatically fetch the correct on based on the name. We could have something like _dm for dark mode that would get the _dm version of an image if it exists. We would need to override the Resources class, don't know how easy it would be, since it's auto-generated. Would also need to add support for Plugin to supply their own dark mode images.

- Why do you use a Func for UseDarkModeSetting? Why not set it directly or just pass it when calling SetDarkMode?

I will check back later to check a little more.

gman-comicdb commented:
> Is there a way why you are loading the uxtheme library using LoadLibraryEx + GetProcAddress. 

No - I don't know any better and that was the example I could find to achieve the goal. Changing.  
<br/>

> If possible when using constant like you calling 5644 on SendMessage, if we can use a named constant instead of the value directly.

Yes - following the existing code conventions, this should go in the `Native` class as constant [`BCM_SETSHIELD`](https://github.com/dahall/Vanara/blob/bf57a670f440fb65671b8a89ecbc9135687a2b6c/PInvoke/User32/WinUser.Button.cs#L564). IMO ideally all the native code would live in the `cYo.Common.Win32` namespace (#todo-one-day).
<br/>

> I see that you are using 5644 to set the UAC Shield icon, there is already an helper method for that you can call.

Funny story... whilst experimenting with `uxtheme`, only one button was themed correctly... can you guess which one? Digging into it led me to that helper method... then _all_ the buttons had shields  :sweat_smile:.
I'm pretty sure the current code is incorrectly calling "set shield off" for all the buttons, with the native OS theming being a side-effect. Should use the correct parameter based on button type.
<br/>

> We should avoid hard coding values, even if it's for ease of refactoring. Things like your DarkColors Dictionary should probably be using a Color Enum instead of a string. When you call it in the SetDarkMode couldn't you just pass the Color directly than the name instead?

Yes - the Dictionary was just because it was a datatype I was more familiar with, it should be a better datatype. I'm not sure how you mean it should be used in `SetDarkMode` but I agree with the sentiment that there's a better way to handle that loop.
<br/>

> Same with hard coding button names like btnVTagsHelp or btAssociateExtensions directly in an unrelated helper file, in an other namespace.

O yh, this is a horrible piece of code. It's there to get the desired UI result but it definitely needs to change. Maybe `if not transparent` for `btnVTagsHelp`, and hopefully the `btAssociateExtensions` exception goes away once I find a way to call "theme button" instead of "set shield off".
<br/>

> Why do you use a Func for UseDarkModeSetting? Why not set it directly or just pass it when calling SetDarkMode?

Good point. I was so focussed on trying to access `UseDarkMode` directly from `ColorSchemes` I didn't even think of that - changed.
<br/>

> I don't know what your idea is for images, the reference I used (and what you seem to have added) is just invert the colors. It might give weird results. we probably could just add alternative images that automatically fetch the correct one based on the name.

It does give weird results, especially with blue becoming yellow. The inversion code actually came late; initially I added alternative images and just modified `Resources.Designer` so I could see what it looked like. The number of images that _need_ to be changed is relatively small; ~25/200. I want to test a brightness inversion that keeps the colors the same (`RGB` -> `HSL` -> invert `L` -> `RGB`).
<br>

> The solution is probably to add some method in your FormEx in virtual that can be overridden. Then in the specific dialog that requires some specific changes just override that method and call that method instead in your SetColorScheme.

That would be pretty cool - I have to read up on virtual methods. `FormEx` came last and it's intentionally light as I was weary of unintended side-effects.
<br/>

> You would have access to all the controls directly and would be able to use all the helper method from the other projects.

So far the biggest use-case for this has been using the custom methods in the extended controls and afaik I proper way to do this would be to inherit from the extended class. E.g. use `ComboBoxEx.GetTextHandle()` on a `ComboBox` control.
Do you see other helper methods that could be used to improve things?
**Side note** - unless there's a reason not to, I think the extended controls should be used instead of the base controls throughout the code base. Currently I'm duplicating code instead of doing that to minimise non-Dark Mode impact.

---

Thanks for your notes, they're really helpful. Going to make suggested changes and continue (`BCM_SETSHIELD` -> ??, `CheckedListBox`, `RichTextBox`, hardcoded color-relating things)

maforget commented:
So If I understand correctly, we are patching the system colors (in `ColorSchemeController`), modifying the `ForeColor`/`BackColor` from the Control (in `ColorSchemeExtensions`) & also are using the uxtheme Windows API (in UXTheme). So all 3 are required? You are also using DarkMode_Explorer & DWMWA_USE_IMMERSIVE_DARK_MODE. Unsure of what that does, I haven't read the documentation.

- What I meant about in `SetDarkMode` you are finding the value for `knownColor` and using it, why not use `knownColor` when using `GetDarkSystemColor` instead of using the property name.
- In `ColorSchemeExtensions` you are using SendMessage to set the `EM_SETCUEBANNER`. It already exists in `ComboBoxEx.SetPromptText`, we could probably make that a `ComboBoxExtension` method like `TextBoxExtensions.SetCueText` and use it in both place using that.
- Why not call `UXTheme` from `ColorSchemeExtensions` directly. Why do we need to cycle through all controls twice? In addition, you have native calls to uxtheme in both `UXTheme` & `ColorSchemeExtensions` class. We keep them all in place, at least the native calls.
- If you call recursively on the control to call uxtheme and on all it children, why do we also need to call them in controls like TextBoxEx ?
- `ApplyDarkThemeRecursive(child, true)` gets called twice inside the foreach. It's called before the `child.Controls.Count` check, so any check after that doesn't matter. The `child.Controls.Count` is kinda cancelled out when it calls the `parent.Controls` in the for loop.
- Check if any of the controls are null like when using `parent?.Controls`
- On that note would `control.IsHandleCreated` give an error if the control is null? Wouldn't `IsHandleCreated` be unavailable to check when the control is null? Seriously don't know. Checking for handles isn't really something that is widely used in C# usually, that seems more like a C++ thing.
- For virtual, when you call `ColorSchemeExtensions.SetColorScheme`, you can check if it's a `FormEx` type and if so then can call a particular method. In `FormEx` you create an empty implementation in virtual so that you don't need to implement it in all the forms and is what is called when not required. Then in those forms that require it you override that method, when overridden it will call that override, instead of the default implementation.
- You could also use an Interface that you assign to the `FormEx`. For ideas check for `FindActiveService`/`QueryService`/`AddService`. They use reflection to automatically finds an object that implements that Interface and call that function. It used a lot in this program. As an example, check `IRefreshDisplay`. It's registered using `CommandMapper.AddService` in a form and calling it via the associated shortcut (or the menus) calls it on any of it's control that is in focus. Be it the tree view or the main control that contains the thumbnails (`ComicBrowserControl`).

maforget commented:
- That code that requires WIndows 11 is that for the ComboBox? They look good on Windows 11. What does it look like in Windows 10? Are they all white, like using my branch?<br/><br/>I've found documentation that says it's build 22000, why use 18985 then, is that for something else?
> DWMWA_USE_IMMERSIVE_DARK_MODE
> Use with [DwmSetWindowAttribute](https://learn.microsoft.com/en-us/windows/win32/api/dwmapi/nf-dwmapi-dwmsetwindowattribute). Allows the window frame for this window to be drawn in dark mode colors when the dark mode system setting is enabled. For compatibility reasons, all windows default to light mode regardless of the system setting. The pvAttribute parameter points to a value of type BOOL. TRUE to honor dark mode for the window, FALSE to always use light mode.
> 
> This value is supported starting with Windows 11 Build 22000.

- In my branch the tabs render correctly because and yours don't. It's because I set `Application.VisualStyleState = System.Windows.Forms.VisualStyles.VisualStyleState.NonClientAreaEnabled;`. Don't remember what were the implications.

- For DPI scaling, what I know know is that since Framework 4.7 the scaling is better. It requires a form `AutoScaleMode` to be set to `AutoScaleMode.Dpi`. I didn't change them, because it is hard to test with my setup using a simple 1080p monitor. But it requires that `Application.EnableVisualStyles()` be called, so we can't remove it. Don't remember why I had to disable it.

gman-comicdb commented:
> So all 3 are required?

Yes but also no.
- `ColorSchemeExtensions.SetDarkMode()` - Overriding `SystemColors`
  - (partial) sledgehammer approach.
  - Blind and doesn't care about how colors interface with each other
  - Not always respected (because custom control? because `KnownColor`/`ProfessionalColor` that hasn't been modified is being used? because WinForms feature?)
- `UXTheme` - `uxtheme.dll` wrapper
  - This is actual Windows OS theming, not color replacement
  - Includes things like brushes, renders, borders, (maybe animations?)
  - Some things like scrollbars can't be themed any other way
  - The number of controls supported and classes varies with OS, and it's undocumented
    - Right now it's very much in a state of "works on my machine"; adding support for other OS versions is needed
    - `DarkMode_Explorer` is a control theme class. Trying to find out which controls have a theme class, and what that theme class is on different OS versions is the challenge. 
  - Ideally we could call `UXTheme.ApplyDarkThemeToWindow(MainForm)` and call it a day - maybe when .NET 10-GA/11 - 10 still has [DarkMode issues](https://github.com/dotnet/winforms/issues?q=state%3Aopen%20label%3A%22area-DarkMode%22)
    <sub>I haven't verified that `uxtheme` is still the theming method but afaik it is</sub>
- `ColorSchemeExtensions.SetColorScheme()`
  - Would be ideal if it didn't exist
  - It's there for fine-tuning and tweaking -precision scissors, compared to the other two sledgehammer functions
  - Handles things not covered by `SetDarkMode()` - `BackColor`,`ForeColor` but also borders.
    - hopefully this can be done in another way
  - Give the app a modern look - darker background/outer elements, lighter "inner elements"
    - end goal should be small list of elements and only setting background
  - Preps controls for `UXTheme` - properties such as `BorderStyle` and `FlatStyle` affect whether controls are themed or not
  - the `ThemeControl` methods are the latest additions and don't belong there
<br>

> Why not call `UXTheme` from `ColorSchemeExtensions` directly. Why do we need to cycle through all controls twice?

I needed to split them for some reason but that is what needs to happen. Cycling should happen once, and even then ideally selectively.
<br>

I'm still moving things around and trying to find the best place to do things.
Here are things I know need to happen:
- `UXTheme` needs to be pruned to just the `Win32` calls. In it's current namespace there's no flexibility to - for example - handle extended types differently.
- `SetDarkMode` -> `Initialize...`, to better reflect functionality
- `SetColorScheme` -> `SetDark...`
    - Calls three methods
        - `SetDarkColor`
        - `Set<whatever-uxtheme-needs-to-theme-a-control>`
        - `ApplyTheme` (what UXTheme is currently doing)
    - A control may need all or none of these methods applied
    - `ApplyTheme` needs to go from a sledgehammer to being more precise, especially on recent OS versions
    - `ThemeControl` methods may not be needed
<br>


---

> What I meant about in SetDarkMode...

Fair point - probably just got tripped up by C# types. 
<br>

> In ColorSchemeExtensions you are using SendMessage to set the `EM_SETCUEBANNER`

I shouldn't be doing that - removed. I agree with the centralization - IMO the native user32 calls belong in `Win32`. 
<br>

> Check if any of the controls are null like when using `parent?.Controls`
> On that note would `control.IsHandleCreated` give an error if the control is null?

More checks definitely needed... because I'm looping over the controls, would have thought the controls couldn't be null but I'm not sure.
<br>

---

> That code that requires WIndows 11 is that for the ComboBox? They look good on Windows 11. What does it look like in Windows 10? Are they all white, like using my branch?

I think the one that's committed is for `ComboBox` but it's for a lot of things - the other big one is `ListView`.
They look fine on Windows 10. But for example for `ListView` you can either theme the list OR the scrollbar; because the scrollbar **colors** can't be set any other way that's what's being themed. The list has the correct **colors** but isn't themed (will clarify)
<br>

> I've found documentation that says it's build 22000, why use 18985 then, is that for something else?

Eh, I wouldn't trust that. Suspect build 22000 is when Microsoft acknowledged it existed because that's when it started to be fit-for-purpose. In earlier builds it sets a dark title bar... _maybe_ more. (so IMMERSIVE... seriously their DarkMode support is a shit-show)
<br>

> In my branch the tabs render correctly and yours don't. It's because I set Application.VisualStyleState = System.Windows.Forms.VisualStyles.VisualStyleState.NonClientAreaEnabled;. Don't remember what were the implications.
> But it requires that Application.EnableVisualStyles() be called, so we can't remove it. Don't remember why I had to disable it.

In both cases, I think what you're doing is disabling the custom paint methods. E.g. [rendering support check](https://github.com/maforget/ComicRackCE/blob/9231547abd8ee73e7f6739be0ac02a07d249179a/cYo.Common.Windows/Forms/TabBar.cs#L1269) for MainView tabs (removing = black background, but still grey border).
I've got `EnableVisualStyles()` enabled; working through the custom overrides
<br>

> For DPI scaling, what I know know is that since Framework 4.7 the scaling is better.

My concern is doing things like passing a hardcoded number of pixels in `Paint` methods - I'm sure there's a built-in solution for it, just need to remember to handle it. (there's also `FormUtility.ScaleDpi[XY]`)
